### PR TITLE
test uint256 for issue #31

### PR DESCRIPTION
**Pull request by devtodev**
I will keep making the tests for uint256 helper.
If anybody wants make some test cases, tell me in order to split the job

### CODE REVIEW

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -80,11 +81,18 @@ target_link_libraries(
     helpers
     trezor-crypto
     )
-
+target_link_libraries(
+    test_uint256
+    ${GTEST_BOTH_LIBRARIES}
+    ${CMAKE_THREAD_LIBS_INIT}
+    eth
```
Comment (pcppcp) [inline tests/CMakeLists.txt]: `eth` is probably not required.

File: tests/test_uint256.cpp [hunk 2]
```diff
@@ -0,0 +1,779 @@
...
+
+// TODO: Check if the logic implemented in writeu128BE is correct
+TEST(TEST_UINT256, WRITEU128BE)
+{
+	uint128_t number;
+	uint8_t buffer;
+
+	UPPER(number) = 0x0;
+	LOWER(number) = 0x123456789abcde;
+	writeu128BE(&number, &buffer);
```
Comment (pcppcp) [inline tests/test_uint256.cpp]: the buffer here is too short (should be at least `sizeof(uint128_t)`), you are writing beyond its boundary into an invalid memory space.

File: tests/test_uint256.cpp [hunk 3]
```diff
@@ -0,0 +1,779 @@
...
+TEST(TEST_UINT256, SET256_UINT64)
+{
+	uint256_t n256;
+	uint64_t n64 = 123456;
+	UPPER(LOWER(n256)) = 1;
+	UPPER(UPPER(n256)) = 2;
+	LOWER(LOWER(n256)) = 3;
+	LOWER(UPPER(n256)) = 4;
+	set256_uint64(&n256, n64);
+	ASSERT_EQ(UPPER(UPPER(n256)), 0);
```
Comment (pcppcp) [inline tests/test_uint256.cpp]: use `equal256()` here, no need to use macros

File: tests/test_uint256.cpp [hunk 4]
```diff
@@ -0,0 +1,779 @@
...
+
+TEST(TEST_UINT256, COPY256)
+{
+	uint256_t number, result;
+	UPPER(LOWER(number)) = 1;
+	UPPER(UPPER(number)) = 2;
+	LOWER(LOWER(number)) = 3;
+	LOWER(UPPER(number)) = 4;
+	copy256(&result, &number);
+	ASSERT_EQ(UPPER(UPPER(number)), UPPER(UPPER(result)));
```
Comment (pcppcp) [inline tests/test_uint256.cpp]: `equal256()` here

File: tests/test_uint256.cpp [hunk 5]
```diff
@@ -0,0 +1,779 @@
...
+	ASSERT_EQ(LOWER(LOWER(number)), LOWER(LOWER(result)));
+}
+
+TEST(TEST_UINT256, CLEAR128)
+{
+	uint128_t number;
+	UPPER(number) = 0;
+	LOWER(number) = 1;
+	clear128(&number);
+	ASSERT_EQ(LOWER(number), 0);
```
Comment (pcppcp) [inline tests/test_uint256.cpp]: `zero128()`

File: tests/CMakeLists.txt (hunk 1 above)
Comment (devtodev) [reply tests/CMakeLists.txt]: Yes, you are right.

File: tests/test_uint256.cpp (hunk 2 above)
Comment (devtodev) [reply tests/test_uint256.cpp]: Calling  _writeu128BE(&number, &buffer)_ with this values:
    ```number (LOWER 123456 UPPER 89abcd) BUFFER: 0 ```
return
    ```number (LOWER: 123456 UPPER: 0) BUFFER: 0 ```
Is the correct reply for writeu128be?
Or should be one of this:
```  number (LOWER cdab89 UPPER 563412) BUFFER: 0 ``` 
```  number (LOWER 563412 UPPER cdab89) BUFFER: 0  ```

Comment (devtodev) [reply tests/test_uint256.cpp]: BTW, uint128_t is composed by two uint64_t which is unsigned long int with limit of 4,294,967,295 decimal or in hexa FFFFFFFFFFFFFFFF, we can check it running this test
```
	UPPER(n1) = 0x1;
	LOWER(n1) = 0;
	UPPER(n2) = 0;
	LOWER(n2) = 0xFFFFFFFFFFFFFFFF;
	minus128(&n1, &n2, &result);
	ASSERT_EQ(UPPER(result), 0);
	ASSERT_EQ(LOWER(result), 1);
```
So LOWER(number) = 0x123456789abcde; its into their boundary. Right?

**Comment by lorepieri8** [comment]
@pcppcp this should be ready to be merged

**Comment by pcppcp** [comment]
looks good but there's a merge conflict and it needs to be rebased with current master

**Comment by devtodev** [comment]
Rebase ready

File: tests/test_uint256.cpp (hunk 2 above)
Comment (pcppcp) [reply tests/test_uint256.cpp]: this still needs fixing: `writeu128BE` encodes a big endian uint128 into a buffer. The buffer length must be defined as `uint8_t buffer[sizeof(uint128_t)]`

**Pull request closed without merging**